### ISSUE TITLE
Upload the data collected by UDC

### ISSUE THREAD

**Issue by reprogrammer** [vazexqi/CodingSpectator#55]
To be able to compare the behavior of users to the behavior of those submitting their UDC logs to eclipse.org, we need to make sure that UDC collects data from our users and our plugin uploads them to our server periodically.

**Comment by reprogrammer** [comment]
Resolved in 5968e5484b390a50231de5aa8b5545df7ae7470a.

**Comment by reprogrammer** [comment]
UDC shows the user a wizard before uploading its data. Thus, it gives the user another opportunity to cancel the data upload. Besides, by default, it uploads data every five days. This might not cause any problem as long as we ask our participants to accept UDC wizards. But, if it turns out to be a problem we should look into UDC to find out how can we copy the UDC files even though the user cancels its wizards. Another alternative is to disable the original UDC upload altogether and replace it by our file copy. This will allow us to suppress UDC wizards. Besides, we will no longer have to create separate directories in CodingSpectator for every upload because UDC won't empty its directory anymore.

**Comment by vazexqi** [comment]
First approach seems better to me: just hook into performCancel, performFinish, etc. Downside is that we have to rely on the initial interval that is set.

Second approach seems _unexpected_ to me. By _unexpected_ I mean that is suppresses UDC wizards which could surprise the users who are expecting the UDC to collect their data and work as before.

I prefer a method that allows UDC to work as closely as originally designed as possible. Also which method will allow UDC to work as intended once the users uninstall our plug-ins? I am thinking that both should be OK since we are only copying the UDC data and never emptying them.

**Comment by reprogrammer** [comment]
Like you, I prefer a method that doesn't mess with the mission of UDC. One issue with UDC is that it lets the user turn off UDC. Do you think we should disable this feature?

**Comment by reprogrammer** [comment]
I suggest that we change UDC dialogs and preferences such that it no longer allows the user to disable the UDC data collection forever. This means the users could potentially ban data upload to the UDC server but not our repository. 

Whenever UDC tries to upload something to the UDC server, it should prompt the user for his/her permission. Since the user might never accept data upload to the UDC server, we have to copy the UDC data into our repository when some other events are triggered. Examples of such events are cancellations of UDC upload dialog and Eclipse startups. Since copying UDC files to the watched directory is usually fast and doesn't require user intervention, we could do it frequently.

One technical issue with copying UDC at times other than UDC upload events is that we might end up copying the same UDC files to the CodingSpectator watched directory several times. Therefore, if we use the simple timestamp directories for organizing UDC data in the watched directory, we'll end up copying the same UDC file in different subdirectories of the watched directory. One way to prevent this duplication is to copy the UDC files to the same watched path until the next UDC upload. Upon the next UDC upload, we create another timestamp directory that becomes the new destination of UDC files. At anytime, we can determine the destination of UDC files in the watched directory by finding the subdirectory with the largest timestamp.

**Comment by vazexqi** [comment]
Initial attempt to always capture (but also with the option of not uploading) is in 12a9f9438760d1bb51286ee766c23fe4b077b6db

**Comment by reprogrammer** [comment]
Resolved in 901e22a952c79eccb0dd66a735b62b494d83276b.

**Comment by reprogrammer** [comment]
What happens if CodingSpectator tries to both transfer UDC data into the watched directory and upload the watched directory to the Subversion repository at the startup event? What happens if these two operations overlap?

**Comment by vazexqi** [comment]
Probably very bad and unpredictable behavior depending on the OS and file system.

Can we just make use of the listener that we have for CodingSpectator just like what Stas has done?

**Comment by reprogrammer** [comment]
Stas had to buffer his data during a data upload. But, I think it would be easier for us to just avoid the overlap between CodingSpectator data upload and UDC transfer. We could either transfer the UDC data after CondigSpectator data upload or upload CodingSpectator data after the UDC transfer.

I believe it's easier to transfer the UDC data after CodingSpectator upload is finished because we already have an extension point to get notified of the completion of data upload. What do you think?

**Comment by reprogrammer** [comment]
An alternative is to acquire the Subversion lock of the watched directory before transferring UDC data into the watched directory. It doesn't even need to be a blocking operation. We could just check if the lock can be acquired. If the lock is available, we'll acquire the lock and transfer the UDC data. Otherwise, we just postpone the UDC data transfer. This approach would enable us to transfer data at every startup not just every time CodingSpectator uploads its data.

**Comment by reprogrammer** [comment]
As Nick said, we have to be careful about using Subversion locks. Because, CodingSpectator sometimes releases the lock forcefully (See issue #113).

**Comment by reprogrammer** [comment]
How about using some other type of lock, e.g. file system lock?

Let's call anyone who writes into the watched directory a _writer_. We'd like to allow multiple writers that write into disjoint subdirectories to work concurrently. But, we don't want the uploader and some writers to interfere. We can implement this protocol using file system locks.

**Comment by reprogrammer** [comment]
I prefer a locking mechanism than just listening to the finish upload event because of stronger safety guarantees. We cannot rely on upload events because there is no guarantee that another upload won't occur just after the first one. In other words, if a writer begins writing as soon as it gets notified of the finish event of the uploader, it might end up interfering with the next upload event. Even though two consecutive uploads are not probable, relying on upload events isn't as safe as using a locking protocol.

**Comment by vazexqi** [comment]
Are you planning to use [FileLock]("http://download.oracle.com/javase/6/docs/api/java/nio/channels/FileLock.html")?

There might be some issues with using it for our case from reading the Javadoc and also from this discussion on [stackoverflow]("http://stackoverflow.com/questions/2479222/problem-with-java-file-locking-mechanism-filelock-etc"):
1. The behavior is not absolute across OS
2. It does not prevent multiple threads from within the single JVM from reading/writing to it (more serious problem for our needs). 

Would it be better to create a singleton object to writing to our _transfer_ directory? That would be the synchronization point?

**Comment by reprogrammer** [comment]
Thanks for pointing me to the problems of FileLock. I suggest to use a single instance of [ReentrantReadWriteLock](http://download.oracle.com/javase/6/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html) to synchronize writes and uploads. Every data writer will acquire a `readLock` and the upload acquires the `writeLock`. This allows multiple data writers to run concurrently but doesn't allow the data writers to interfere with the uploader.

**Comment by vazexqi** [comment]
Which object will have the ReentrantReadWriteLock that others will try to acquire? That is which plug-in has that lock? The plug-in that has that lock will require all other plug-ins to depend on it either:
1. statically through the MANIFEST.MF
2. dynamically through extension points

Will this also be a change that will affect all existing plug-ins that read/write to the ltk directory? Will this also include LTK core itself or do we not have to worry about that?

**Comment by reprogrammer** [comment]
b28e48c41b270726c715d7c84146f5e8c609522e synchronizes the CodingSpectator submitter and the transfer of UDC data to the CodingSpectator's watched directory using the monitor.core.submitter extension point.

**Comment by reprogrammer** [comment]
It isn't sufficient for the submitter to acquire the locks during `preSubmit` only, because this design doesn't eliminate all unwanted interleavings. For example, if `preSubmit` gets executed before the method `UploadManager.startTransferToCodingSpectator()`, `preSubmit` will acquire and release the locks and the submitter will begin the submission immediately. Meanwhile, `UploadManager.startTransferToCodingSpectator()` could get executed in parallel with the submission. Thus, the copying of UDC data could overlap with the submission of the watched directory.

I suggest that we change the implementation of `preSubmit` in `UploadManager` such that it only acquires the two locks and `postSubmit` releases the two locks.

**Comment by Wanderer777** [comment]
Indeed, the current approach guarantees a particular order (i.e. no interleaving) only when `transferToCodingSpectator` is called before `preSubmit`. So, basically the uploader will wait for the transfer to complete before proceeding, while the transfer will not wait for the uploader to complete.

I agree that acquiring an explicit lock in `preSubmit` and releasing it in `postSubmit` would solve this problem (the same lock should be acquired at the beginning of `transferToCodingSpectator` and released at its end), because this way, both the transfer and the uploader will wait for each other. Though I am not sure why you would need two locks? It seems that a single explicit lock should be enough.

As about CodingTracker, this problem is not relevant to it. It can work in parallel with the submission (i.e. it does not need to wait till it is over in order to start logging), because it switches the log files. So, the only requirement is that this switching is done atomically and can not interleave with the submission, which is achieved by synchronizing the methods, which do the switching (as well as the method that does the logging). This way, the uploader will never proceed before the switching is over, while the CodingTracker's logger can proceed in parallel with the submission because it can never write to the file that is submitted (as it is switched to write to a temporary file during the submission).

**Comment by reprogrammer** [comment]
Thanks for your detailed comment. You're right that a single lock is enough to make the submission and the UDC data transfer mutually exclusive. And, as you explained, CodingTracker is safe without a lock that spans `preSubmit` and `postSubmit` because it switches log files.

I applied this change in 3e42f2afc3f97b33977212f7cdb16c02df4a8c8f.

**Comment by reprogrammer** [comment]
Merged the branch `usage-data-collector` into `master` in 1b7ef68a20883df15f5ce72a28bfba0be9b2bd63.
